docs/spec_draft.md

```diff
@@ -300,6 +316,154 @@ the IEEE-754 specification. For boolean element type, the behavior is same as
+   * The formation of `groups` can be demonstrated as follows:
+     ```python
+     if len(replica_groups) == 0:
+       replica_groups = all_partions_ids
+
+     for g in replica_groups:
+       for pid in all_partions_ids:
+         for rid in `g`:
+           group += ei(rid, pid)
+         groups += group
+     return groups
+     ```
+    * An examples assuming P = 2: `replica_groups`: `[[0, 2], [1, 3]]` and the
+      `groups` formed: `{ei(0, 0), ei(2, 0)}, {ei(0, 1), ei(2, 1)},  {ei(1, 0), ei(3, 0)}, and {ei(1, 1), ei(3, 1)}`.
+
+2. If `channel_handle` is not provided and `use_global_device_ids` $=$ `true`
+   * Invalid configuration.
+
+3. If `channel_handle` is provided and `use_global_device_ids` = false
+   * Number of groups formed: G.
+   * Size of each group: S * P.
+   * Interpretation of `replica_groups`: contains replica ids.
+   * The formation of `groups` can be demonstrated as follows:
+     ```python
+     if len(replica_groups) == 0:
+       replica_groups = all_partions_ids
+
+     for g in replica_groups:
+       for rid in g:
+         for pid in all_partions_ids:
+           group += ei(rid, pid)
+       groups += group
+     return groups
+     ```
+    * An example assuming P = 2: `replica_groups`: `[[0, 2], [1, 3]]` and the
+      `groups` formed: `{ei(0, 0), ei(0, 1), ei(2, 0), ei(2, 1)} and {ei(1, 0), ei(1, 1), ei(3, 0), ei(3, 1)}`.
+
+3. If `channel_handle` is provided and `use_global_device_ids` $=$ `true`
+   * Number of groups formed: G.
+   * Size of each group: S.
+   * Interpretation of `replica_groups`: contains global device ids.
+   * The formation of `groups` can be demonstrated as follows:
+     ```python
+     for g in replica_groups:
+       for global_device_id in g:
+           group += ei(global_device_id / P, global_device_id % P)
+       groups += group
+     return groups
+     ```
+    * An example assuming P = 2: `replica_groups`: `[[4, 5], [6, 7]]` and the
+      `groups` formed: `{ei(2, 0), ei(2, 1)} and {ei(3, 0), ei(3, 1)}`.
+
+For each `group` $\in$ `groups`, the `result`, to be visible by each member of
+`group`, can be expressed as follows:
+
+```python
+result[i0, i1, ..., iR-1] = reduce(
+      inputs= operand[i0, i1, ..., iR-1] for each `ei(rid, pid)` in group,
+      init_values=[I, such that `computation` and `I` form a monoid],
+      dimensions=[i for i in (0, R)],
+      body=computation)
+```
+The order of reductions is implementation-defined.
+
+### Inputs
+
+| Name                    | Type                                                                                                                                         |
+|-------------------------|----------------------------------------------------------------------------------------------------------------------------------------------|
+| `operand`               | tensor of any supported type                                                                                                                 |
```

Comment:
`all_gather_dim` is missing here

Reply:
I had a type to include `all_gather_dims` in this op. Revised the constraints.